services/src/main/java/edu/unc/lib/dl/cdr/services/processing/XMLExportJob.java

```diff
@@ -0,0 +1,194 @@
+import javax.mail.internet.MimeMessage;
+
+import org.jdom2.Document;
+import org.jdom2.Element;
+import org.jdom2.input.SAXBuilder;
+import org.jdom2.output.Format;
+import org.jdom2.output.XMLOutputter;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.mail.javamail.JavaMailSender;
+import org.springframework.mail.javamail.MimeMessageHelper;
+
+import edu.unc.lib.dl.acl.service.AccessControlService;
+import edu.unc.lib.dl.acl.util.AccessGroupSet;
+import edu.unc.lib.dl.acl.util.GroupsThreadStore;
+import edu.unc.lib.dl.acl.util.Permission;
+import edu.unc.lib.dl.cdr.services.rest.modify.ExportXMLController.XMLExportRequest;
+import edu.unc.lib.dl.fcrepo4.BinaryObject;
+import edu.unc.lib.dl.fcrepo4.ContentObject;
+import edu.unc.lib.dl.fcrepo4.PIDs;
+import edu.unc.lib.dl.fcrepo4.RepositoryObjectLoader;
+import edu.unc.lib.dl.fedora.PID;
+
+
+    /**
+     * Job that performs the work of retrieving metadata documents and compiling them into the export document.
+     *
+     * @author bbpennel
+     * @author harring
+     */
+public class XMLExportJob implements Runnable {
+    private static final Logger log = LoggerFactory.getLogger(XMLExportJob.class);
+
+    @Autowired
+    private JavaMailSender mailSender;
+    @Autowired
+    private AccessControlService aclService;
+    @Autowired
+    private RepositoryObjectLoader repoObjLoader;
+
+    private final int BUFFER_SIZE = 2048;
+    private final Charset utf8 = Charset.forName("UTF-8");
+    private final String separator = System.getProperty("line.separator");
+    private final byte[] separatorBytes = System.getProperty("line.separator").getBytes();
+    private final byte[] exportHeaderBytes = ("<?xml version=\"1.0\" encoding=\"utf-8\"?>" + separator
+            + "<bulkMetadata>" + separator).getBytes(utf8);
+
+    private final String user;
+    private final AccessGroupSet groups;
+    private final XMLExportRequest request;
+
+    public XMLExportJob(String user, AccessGroupSet groups, XMLExportRequest request) {
+        this.user = user;
+        this.groups = groups;
+        this.request = request;
+    }
+
+    @Override
+    public void run() {
+        long startTime = System.currentTimeMillis();
+
+        try {
+            GroupsThreadStore.storeGroups(groups);
+            GroupsThreadStore.storeUsername(user);
+
+            File mdExportFile = File.createTempFile("xml_export", ".xml");
+
+            try (FileOutputStream xfop = new FileOutputStream(mdExportFile)) {
+                xfop.write(exportHeaderBytes);
+
+                XMLOutputter xmlOutput = new XMLOutputter(Format.getRawFormat());
+                for (String pidString : request.getPids()) {
+                    PID pid = PIDs.get(pidString);
+                    ContentObject obj = (ContentObject) repoObjLoader.getRepositoryObject(pid);
+                    BinaryObject mods = obj.getMODS();
```

Comment:
Two suggestions here:
1) the `<object pid="blah"></object>` tag needs to be present even if there is no MODS for the object.
2) I would suggest verifying the user has access before pulling the MODS or adding the `<object>` tag.  It is a much cheaper operation than pulling the MODS binary from disk.